src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMarker.cs

```diff
@@ -70,27 +71,33 @@ internal void MarkTypeSystemEntity(in MessageOrigin origin, TypeSystemEntity ent
                     MarkEvent(origin, @event, reason);
                     break;
                     // case InterfaceImplementation
-                    //  Nothing to do currently as Native AOT will presere all interfaces on a preserved type
+                    //  Nothing to do currently as Native AOT will preserve all interfaces on a preserved type
             }
         }
 
         internal bool TryResolveTypeNameAndMark(string typeName, in DiagnosticContext diagnosticContext, bool needsAssemblyName, string reason, [NotNullWhen(true)] out TypeDesc? type)
         {
             ModuleDesc? callingModule = ((diagnosticContext.Origin.MemberDefinition as MethodDesc)?.OwningType as MetadataType)?.Module;
 
-            // NativeAOT doesn't have a fully capable type name resolver yet
-            // Once this is implemented don't forget to wire up marking of type forwards which are used in generic parameters
-            if (!DependencyAnalysis.ReflectionMethodBodyScanner.ResolveType(typeName, callingModule, diagnosticContext.Origin.MemberDefinition!.Context, out TypeDesc foundType, out ModuleDesc referenceModule))
+            List<ModuleDesc> referencedModules = new();
+            TypeDesc foundType = System.Reflection.TypeNameParser.ResolveType(typeName, callingModule, diagnosticContext.Origin.MemberDefinition!.Context, referencedModules);
+            if (foundType == null)
             {
+                if (needsAssemblyName)
+                    diagnosticContext.AddDiagnostic(DiagnosticId.TypeWasNotFoundInAssemblyNorBaseLibrary, typeName);
```

Comment:
I am wondering about usefulness of this warning. `Type.GetType` is often used to check whether the type exists.

Why do we want to warn when the type does not exist? Assuming there are no bugs, the type won't be found at runtime either. 

Are there interesting cases where the type is not found at compile time, but somehow found at runtime?

Reply:
There's some discussion on this in https://github.com/dotnet/linker/pull/1920, but it didn't seem to be "closed" in any way. I agree the value of this is low (it only triggers for data flow marking, so if somebody as a `string` variable annotated  with DAM and the constant value passed to that variable is a type name without assembly name and it doesn't resolve).

I'd be fine if we removed the warning completely. There's also a comment about this for the analyzer where we don't have a solution for it either: https://github.com/dotnet/runtime/blob/f561a058b13523c4b045803c9876e823b4cfe0fc/src/tools/illink/src/ILLink.RoslynAnalyzer/TrimAnalysis/RequireDynamicallyAccessedMembersAction.cs#L28-L30

Reply:
IIRC we only want to warn if:
* this is an assignment to a DAM-annotated location of type System.String, and
* the string wasn't assembly-qualified

We generate the warning to cover this crazy scenario:

Assembly A:

```csharp
Get("SomeType"); // not assembly qualified!
```

Assembly B:

```csharp
static void Get([DAM] string b) => Type.GetType(b);

class SomeType { }
```

We're not going to mark SomeType at the callsite to Get because we don't know what assembly it's coming from (doesn't resolve).

However, at runtime, the Type.GetType call is going to look into the calling assembly.

There would be no other warning and the code would be broken.

Reply:
Oh well - in that case let's keep it - especially since it's already there.
Jan could you please add a comment to the code about the case above?